connect/tunnel_api/tunnel_api.go

```diff
@@ -0,0 +1,785 @@
+package tunnel_api
+
+import (
+	"bytes"
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"fmt"
+	"math/big"
+	"net"
+	"strconv"
+	"strings"
+	"text/template"
+	"time"
+
+	dhparam "github.com/Luzifer/go-dhparam"
+	securityv1 "github.com/openshift/api/security/v1"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	dapi "k8s.io/client-go/discovery"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	openvpnClientConfTemplate = `client
+cipher AES-256-GCM
+dev tun
+proto tcp4
+remote {{ $.Endpoint }} {{ $.Port }} tcp-client
+<ca>
+{{ .CA }}
+</ca>
+<key>
+{{ .Key }}
+</key>
+<cert>
+{{ .Crt }}
+</cert>
+verify-x509-name "C=US, ST=NC, L=RDU, O=Engineering, OU=Crane, CN=Server"
+`
+	openvpnServerConfTemplate = `dh /certs/dh.pem
+cipher AES-256-GCM
+ca /certs/ca.crt
+server 192.168.123.0 255.255.255.0
+dev tun0
+proto tcp4
+port {{ $.Port }}
+keepalive 10 120
+tmp-dir /tmp/openvpn
+cert /certs/server.crt
+key /certs/server.key
+`
+	serviceName   = "openvpn"
+	serviceConfig = "openv1pn-conf"
+	keySize       = 2048
+)
+
+type Tunnel struct {
+	DstClient client.Client
+	DstConfig *rest.Config
+	SrcClient client.Client
+	SrcConfig *rest.Config
+	Options   Options
+}
+
+type Options struct {
+	Namespace   string
+	CACrt       *bytes.Buffer
+	ClientCrt   *bytes.Buffer
+	ClientKey   *bytes.Buffer
+	ServerCrt   *bytes.Buffer
+	ServerKey   *bytes.Buffer
+	RSADHKey    *bytes.Buffer
+	ClientImage string
+	ServerImage string
+	ServerPort  int32
+}
+
+type openvpnConfigData struct {
+	Port     string
+	CA       string
+	Crt      string
+	Key      string
+	Endpoint string
+}
+
+func Openvpn(tunnel Tunnel) error {
+	if tunnel.Options.Namespace == "" {
+		tunnel.Options.Namespace = serviceName
+	}
+
+	if tunnel.Options.ClientImage == "" {
+		tunnel.Options.ClientImage = "quay.io/konveyor/openvpn:latest"
+	}
+
+	if tunnel.Options.ServerImage == "" {
+		tunnel.Options.ServerImage = "quay.io/konveyor/openvpn:latest"
+	}
+	if tunnel.Options.ServerPort == 0 {
+		tunnel.Options.ServerPort = int32(1194)
+	}
+	if tunnel.Options.CACrt == nil {
+		ca, serverCrt, serverKey, clientCrt, clientKey, dh, err := GenOpenvpnSSLCrts()
+		if err != nil {
+			return err
+		}
+		tunnel.Options.CACrt = ca
+		tunnel.Options.ServerCrt = serverCrt
+		tunnel.Options.ServerKey = serverKey
+		tunnel.Options.ClientCrt = clientCrt
+		tunnel.Options.ClientKey = clientKey
+		tunnel.Options.RSADHKey = dh
+	}
+
+	scheme := runtime.NewScheme()
```

Comment:
@alaypatel07 is using a "Register" function, that will register the types to a passed in the scheme, I wonder if we can follow a similar pattern?

Reply:
Just learning, https://github.com/openshift/api/blob/1456ef82ae5df7802a9f7a5f7baeece6fc5c991a/route/v1/register.go#L15 using `routev1.Install(defaultScheme)` might be the right way to add things to controller-runtime scheme. It was called AddToScheme but is now depracated https://github.com/openshift/api/blob/1456ef82ae5df7802a9f7a5f7baeece6fc5c991a/route/v1/register.go#L22

Reply:
`undefined: "k8s.io/api/apps/v1".Install`

Reply:
After talking with @alaypatel07 he is not recommended a different approach at this time. Doing this would require vendoring kubernetes, but @alaypatel07 has some other thoughts on centralizing and exposing this logic.

https://github.com/konveyor/crane-lib/issues/87 is covered to track that work separately.